jerry-core/api/jerry.c

```diff
@@ -6581,6 +6581,171 @@ jerry_get_array_from_container (jerry_value_t value, /**< the container or itera
+ *
+ * @return error - if argument is invalid or operation is unsuccessful or unsupported
+ *                 result of the container operation - otherwise.
+ */
+jerry_value_t
+jerry_container_operation (jerry_container_operation_t operation, /**< container operation */
+                           jerry_value_t container, /**< container */
+                           jerry_value_t *arguments, /**< list of arguments */
+                           uint32_t arguments_number) /**< number of arguments */
+{
+  jerry_assert_api_available ();
+#if JERRY_BUILTIN_CONTAINER
+  if (!ecma_is_value_object (container))
+  {
+    return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("Container is not an object.")));
+  }
+
+  ecma_object_t *obj_p = ecma_get_object_from_value (container);
+
+  if (ecma_get_object_type (obj_p) != ECMA_OBJECT_TYPE_CLASS)
+  {
+    return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG ("Container is not an object.")));
+  }
+  uint16_t type = ((ecma_extended_object_t *) obj_p)->u.cls.u2.container_id;
+  ecma_extended_object_t *container_object_p = ecma_op_container_get_object (container, type);
+
+  if (container_object_p == NULL)
+  {
+    return ecma_create_error_reference_from_context ();
+  }
+
+  const char *incorrect_type_call = ECMA_ERR_MSG ("Operator called on incorrect container type");
+  if (type == LIT_MAGIC_STRING_SET_UL || (type == LIT_MAGIC_STRING_WEAKSET_UL))
+  {
+    if (operation == JERRY_CONTAINER_OP_GET || operation == JERRY_CONTAINER_OP_SET)
+    {
+      return jerry_throw (ecma_raise_type_error (incorrect_type_call));
+    }
+    if (type == LIT_MAGIC_STRING_WEAKSET_UL && operation == JERRY_CONTAINER_OP_CLEAR)
+    {
+      return jerry_throw (ecma_raise_type_error (incorrect_type_call));
+    }
+  }
+
+  if (type == LIT_MAGIC_STRING_MAP_UL || (type == LIT_MAGIC_STRING_WEAKMAP_UL))
+  {
+    if (operation == JERRY_CONTAINER_OP_ADD)
+    {
+      return jerry_throw (ecma_raise_type_error (incorrect_type_call));
+    }
+    if (type == LIT_MAGIC_STRING_WEAKMAP_UL && operation == JERRY_CONTAINER_OP_CLEAR)
+    {
+      return jerry_throw (ecma_raise_type_error (incorrect_type_call));
+    }
+  }
+
+  switch (arguments_number)
+  {
+    case 0:
+    {
+      if (operation != JERRY_CONTAINER_OP_SIZE && operation != JERRY_CONTAINER_OP_CLEAR)
+      {
+        return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG (ecma_error_wrong_args_msg_p)));
+      }
+      break;
+    }
+    case 1:
+    {
+      if (ecma_is_value_error_reference (arguments[0]))
+      {
+        return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG (ecma_error_wrong_args_msg_p)));
+      }
+
+      if (operation == JERRY_CONTAINER_OP_SIZE || operation == JERRY_CONTAINER_OP_CLEAR)
+      {
+        return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG (ecma_error_wrong_args_msg_p)));
+      }
+
+      if (operation == JERRY_CONTAINER_OP_SET)
```

Comment:
This could be combined with the check above. But it would be better if you would check the allowed operations, because that is easier to maintain.